cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/SecureMultiStore.java

```diff
@@ -0,0 +1,267 @@
+      return listMetadataInternal(context, namespace);
+    }, IOException.class);
+  }
+
+  @Override
+  public void create(NamespaceResourceReference ref, R obj)
+      throws AccessException, IOException, IllegalArgumentException, IllegalStateException {
+    Util.validateResourceName(ref);
+    String version = createIndex(ref);
+    writeSecureData(ref, obj, version);
+    TransactionRunners.run(transactionRunner, context -> {
+      if (getMetadataInternal(context, ref).isPresent()) {
+        throw new NamespaceResourceAlreadyExistsException(ref);
+      }
+      writeMetadataInternal(context, ref, obj.getMetadata(), version);
+      updateLiveIndex(context, ref, version);
+    }, IOException.class);
+  }
+
+  @Override
+  public void update(NamespaceResourceReference ref, R obj)
+      throws IOException, IllegalStateException {
+    String version = createIndex(ref);
+    writeSecureData(ref, obj, version);
+    TransactionRunners.run(transactionRunner, context -> {
+      if (!getMetadataInternal(context, ref).isPresent()) {
+        throw new NamespaceResourceNotFoundException(ref);
+      }
+      writeMetadataInternal(context, ref, obj.getMetadata(), version);
+      updateLiveIndex(context, ref, version);
+    }, IOException.class);
+  }
+
+  @Override
+  public void delete(NamespaceResourceReference ref) throws AccessException, IOException,
+      IllegalStateException, NamespaceResourceNotFoundException {
+    Optional<VersionedObject<T>> obj = TransactionRunners.run(transactionRunner, context -> {
+      return getMetadataInternal(context, ref);
+    });
+    if (!obj.isPresent()) {
+      throw new NamespaceResourceNotFoundException(ref);
+    }
+    String secureStoreKey = constructSecureStoreKey(ref, obj.get().getVersion());
+    try {
+      secureStoreManager.delete(NamespaceId.SYSTEM.getNamespace(), secureStoreKey);
+    } catch (Exception e) {
+      // Key may have already been deleted, so just log an error.
+      LOG.warn("Failed to delete key '{}' from system secure store", secureStoreKey, e);
+    }
+    TransactionRunners.run(transactionRunner, context -> {
+      deleteMetadataInternal(context, ref);
+    }, IOException.class);
+  }
+
+  /*
+   * Helper methods
+   */
+
+  private String computeNamespaceSalt(String namespace) {
+    try {
+      MessageDigest messageDigest = MessageDigest.getInstance("SHA-256");
+      return Bytes.toHexString(messageDigest.digest(namespace.getBytes())).substring(0, 8);
+    } catch (NoSuchAlgorithmException e) {
+      // This should never happen.
+      throw new RuntimeException(String.format("Failed to compute salt for namespace '%s'",
+          namespace), e);
+    }
+  }
+
+  private String constructSecureStoreKey(NamespaceResourceReference ref, String version) {
+    return String.format("%s-%s-%s-%s-%s", getResourceType(), ref.getNamespace(),
+        computeNamespaceSalt(ref.getNamespace()), ref.getName(), version);
```

Comment:
why is a salt needed?

Reply:
This was originally needed because namespaces are case-sensitive, but secure keys are not. However, we've moved away from this design entirely, so this is no longer applicable.